File: src/test/java/com/heekng/celloct/service/WorkUpdateRequestServiceTest.java
package com.heekng.celloct.service;

import com.heekng.celloct.dto.WorkUpdateRequestDto;
import com.heekng.celloct.entity.*;
import com.heekng.celloct.repository.*;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;

@SpringBootTest
@Transactional
class WorkUpdateRequestServiceTest {

    @Autowired
    WorkUpdateRequestService workUpdateRequestService;
    @Autowired
    WorkUpdateRequestRepository workUpdateRequestRepository;
    @Autowired
    ShopRepository shopRepository;
    @Autowired
    MemberRepository memberRepository;
    @Autowired
    StaffRepository staffRepository;
    @Autowired
    WorkRepository workRepository;
    @Autowired
    EntityManager em;

    @Test
    @DisplayName("근무시간 수정 요청 추가")
    void addTest() throws Exception {
        //given
        Shop shop = Shop.builder()
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Member member = Member.builder()
                .name("member1")
                .email("test@gmail.com")
                .build();
        memberRepository.save(member);

        Staff staff = Staff.builder()
                .name("staff1")
                .shop(shop)
                .member(member)
                .build();
        staffRepository.save(staff);

        LocalDate workDate = LocalDate.of(2022, 10, 10);
        LocalDateTime startDate = LocalDateTime.of(2022, 10, 10, 5, 10);
        LocalDateTime endDate = LocalDateTime.of(2022, 10, 10, 6, 10);

        Work work = Work.builder()
                .workDate(workDate)
                .startDate(startDate)
                .endDate(endDate)
                .staff(staff)
                .build();
        workRepository.save(work);

        //when
        WorkUpdateRequestDto.AddRequest addRequest = WorkUpdateRequestDto.AddRequest.builder()
                .staffId(staff.getId())
                .workId(work.getId())
                .workDate(workDate)
                .startDate(startDate.minusHours(1))
                .endDate(endDate.plusHours(1))
                .build();
        Long workUpdateRequestId = workUpdateRequestService.addWorkUpdateRequest(addRequest);
        WorkUpdateRequest findWorkUpdateRequest = workUpdateRequestRepository.findById(workUpdateRequestId).get();

        //then
        assertThat(work).isEqualTo(findWorkUpdateRequest.getWork());
    }

    @Test
    @DisplayName("근무시간 수정 요청 수정")
    void updateTest() throws Exception {
        //given
        Shop shop = Shop.builder()
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Member member = Member.builder()
                .name("member1")
                .email("test@gmail.com")
                .build();
        memberRepository.save(member);

        Staff staff = Staff.builder()
                .name("staff1")
                .shop(shop)
                .member(member)
                .build();
        staffRepository.save(staff);

        LocalDate workDate = LocalDate.of(2022, 10, 10);
        LocalDateTime startDate = LocalDateTime.of(2022, 10, 10, 5, 10);
        LocalDateTime endDate = LocalDateTime.of(2022, 10, 10, 6, 10);

        Work work = Work.builder()
                .workDate(workDate)
                .startDate(startDate)
                .endDate(endDate)
                .staff(staff)
                .build();
        workRepository.save(work);

        WorkUpdateRequest workUpdateRequest = WorkUpdateRequest.builder()
                .work(work)
                .workDate(workDate)
                .startDate(startDate.minusHours(1))
                .endDate(startDate.plusHours(1))
                .build();
        workUpdateRequestRepository.save(workUpdateRequest);

        em.flush();
        em.clear();

        //when
        LocalDateTime updateStartDate = workUpdateRequest.getWorkTime().getStartDate().minusHours(1);
        WorkUpdateRequestDto.UpdateRequest updateRequest = WorkUpdateRequestDto.UpdateRequest.builder()
                .workUpdateRequestId(workUpdateRequest.getId())
                .updateDate(workUpdateRequest.getWorkTime().getWorkDate())
                .updateStartDate(updateStartDate)
                .updateEndDate(workUpdateRequest.getWorkTime().getEndDate())
                .build();

        workUpdateRequestService.updateWorkUpdateRequest(updateRequest);
        //then
        assertThat(updateRequest.getUpdateStartDate()).isEqualTo(updateStartDate);
    }

    @Test
    @DisplayName("근무시간 수정 요청 삭제")
    void deleteTest() throws Exception {
        //given
        Shop shop = Shop.builder()
                .name("shop1")
                .build();
        shopRepository.save(shop);

        Member member = Member.builder()
                .name("member1")
                .build();
        memberRepository.save(member);

        Staff staff = Staff.builder()
                .shop(shop)
                .member(member)
                .build();
        staffRepository.save(staff);

        LocalDate workDate = LocalDate.of(2022, 10, 10);
        LocalDateTime startDate = LocalDateTime.of(2022, 10, 10, 5, 10);
        LocalDateTime endDate = LocalDateTime.of(2022, 10, 10, 6, 10);

        Work work = Work.builder()
                .workDate(workDate)
                .startDate(startDate)
                .endDate(endDate)
                .staff(staff)
                .build();
        workRepository.save(work);

        WorkUpdateRequest workUpdateRequest = WorkUpdateRequest.builder()
                .work(work)
                .workDate(workDate)
                .startDate(startDate.minusHours(1))
                .endDate(endDate)
                .build();
        workUpdateRequestRepository.save(workUpdateRequest);

        //when
        workUpdateRequestService.deleteWorkUpdateRequest(workUpdateRequest.getId());
        Optional<WorkUpdateRequest> workUpdateRequestOptional = workUpdateRequestRepository.findById(workUpdateRequest.getId());
        //then
        assertThat(workUpdateRequestOptional).isEmpty();

    }



}

File: src/test/kotlin/com/heekng/celloct/service/WorkUpdateRequestServiceTest.kt
package com.heekng.celloct.service

import com.heekng.celloct.dto.WorkUpdateRequestDto.AddRequest
import com.heekng.celloct.dto.WorkUpdateRequestDto.UpdateRequest
import com.heekng.celloct.entity.*
import com.heekng.celloct.repository.*
import com.heekng.celloct.util.findByIdOrThrow
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.repository.findByIdOrNull
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import javax.persistence.EntityManager

@SpringBootTest
@Transactional
class WorkUpdateRequestServiceTest @Autowired constructor(
    private val workUpdateRequestService: WorkUpdateRequestService,
    private val workUpdateRequestRepository: WorkUpdateRequestRepository,
    private val shopRepository: ShopRepository,
    private val memberRepository: MemberRepository,
    private val staffRepository: StaffRepository,
    private val workRepository: WorkRepository,
    private val em: EntityManager,
) {

    @Test
    @DisplayName("근무시간 수정 요청 추가 테스트")
    fun addTest() {
        //given
        val shop = Shop.fixture(null, "shop", null)
        shopRepository.save(shop)
        val member = Member.fixture("member", "test@gmail.com")
        memberRepository.save(member)
        val staff = Staff.fixture(member, shop, "staff")
        staffRepository.save(staff)
        val workDate = LocalDate.of(2022, 10, 10)
        val startDate = LocalDateTime.of(2022, 10, 10, 5, 10)
        val endDate = LocalDateTime.of(2022, 10, 10, 6, 10)
        val work = Work(
            workTime = WorkTime(
                workDate = workDate,
                startDate = startDate,
                endDate = endDate,
            ) ,
            staff = staff,
        )
        workRepository.save(work)
        //when
        val addRequest = AddRequest(work.id, staff.id, workDate, startDate.minusHours(1), endDate.plusHours(1), null)
        val workUpdateRequestId = workUpdateRequestService.addWorkUpdateRequest(addRequest)
        val findWorkUpdateRequest = workUpdateRequestRepository.findByIdOrThrow(workUpdateRequestId)
        //then
        assertThat(findWorkUpdateRequest.work).isEqualTo(work)
    }

    @Test
    @DisplayName("근무시간 요청 수정")
    fun updateTest() {
        //given
        val shop = Shop.fixture(null, "shop", null)
        shopRepository.save(shop)
        val member = Member.fixture("member", "test@gmail.com")
        memberRepository.save(member)
        val staff = Staff.fixture(member, shop, "staff")
        staffRepository.save(staff)
        val workDate = LocalDate.of(2022, 10, 10)
        val startDate = LocalDateTime.of(2022, 10, 10, 5, 10)
        val endDate = LocalDateTime.of(2022, 10, 10, 6, 10)
        val work = Work(
            workTime = WorkTime(
                workDate = workDate,
                startDate = startDate,
                endDate = endDate,
            ) ,
            staff = staff,
        )
        workRepository.save(work)
        val workUpdateRequest = WorkUpdateRequest.fixture(workDate, startDate.minusHours(1), endDate.plusHours(1), work, null)
        workUpdateRequestRepository.save(workUpdateRequest)
        em.flush()
        em.clear()
        //when
        val updateStartDate = workUpdateRequest.workTime.startDate.minusHours(1)
        val updateRequest = UpdateRequest(
            workUpdateRequest.id,
            workUpdateRequest.workTime.workDate,
            updateStartDate,
            workUpdateRequest.workTime.endDate
        )
        workUpdateRequestService.updateWorkUpdateRequest(updateRequest)
        val findWorkUpdateRequest = workUpdateRequestRepository.findByIdOrThrow(workUpdateRequest.id)
        //then
        assertThat(findWorkUpdateRequest.workTime.startDate).isEqualTo(updateStartDate)
    }

    @Test
    @DisplayName("근무시간 수정 요청 삭제 테스트")
    fun deleteTest() {
        //given
        val shop = Shop.fixture(null, "shop", null)
        shopRepository.save(shop)
        val member = Member.fixture("member", "test@gmail.com")
        memberRepository.save(member)
        val staff = Staff.fixture(member, shop, "staff")
        staffRepository.save(staff)
        val workDate = LocalDate.of(2022, 10, 10)
        val startDate = LocalDateTime.of(2022, 10, 10, 5, 10)
        val endDate = LocalDateTime.of(2022, 10, 10, 6, 10)
        val work = Work(
            workTime = WorkTime(
                workDate = workDate,
                startDate = startDate,
                endDate = endDate,
            ) ,
            staff = staff,
        )
        workRepository.save(work)
        val workUpdateRequest = WorkUpdateRequest.fixture(workDate, startDate.minusHours(1), endDate.plusHours(1), work, null)
        workUpdateRequestRepository.save(workUpdateRequest)
        em.flush()
        em.clear()
        //when
        workUpdateRequestService.deleteWorkUpdateRequest(workUpdateRequest.id)
        val findWorkUpdateRequestOrNull = workUpdateRequestRepository.findByIdOrNull(workUpdateRequest.id)
        //then
        assertThat(findWorkUpdateRequestOrNull).isNull()
    }
}